src/cart/crt_context.c

```diff
@@ -635,39 +700,9 @@ crt_context_destroy(crt_context_t crt_ctx, int force)
 	if (!force && rc && i == CRT_SWIM_FLUSH_ATTEMPTS)
 		D_GOTO(err_unlock, rc);
 
-	rc = d_hash_table_destroy_inplace(&ctx->cc_epi_table, true /* force */);
-	if (rc) {
-		D_ERROR("destroy context (idx %d, force %d), "
-			"d_hash_table_destroy_inplace failed, rc: %d.\n",
-			ctx->cc_idx, force, rc);
-		if (!force)
-			D_GOTO(err_unlock, rc);
-	}
-
-	d_binheap_destroy_inplace(&ctx->cc_bh_timeout);
-
 	D_MUTEX_UNLOCK(&ctx->cc_mutex);
 
-	provider = ctx->cc_hg_ctx.chc_provider;
-
-	rc = crt_hg_ctx_fini(&ctx->cc_hg_ctx);
-	if (rc) {
-		D_ERROR("crt_hg_ctx_fini failed() rc: " DF_RC "\n", DP_RC(rc));
-
-		if (!force)
-			D_GOTO(out, rc);
-	}
-
-	D_RWLOCK_WRLOCK(&crt_gdata.cg_rwlock);
-
-	crt_provider_put_ctx_idx(ctx->cc_primary, provider, ctx->cc_idx);
-	d_list_del(&ctx->cc_link);
-
-	D_RWLOCK_UNLOCK(&crt_gdata.cg_rwlock);
-
-	D_MUTEX_DESTROY(&ctx->cc_mutex);
-	D_DEBUG(DB_TRACE, "destroyed context (idx %d, force %d)\n", ctx->cc_idx, force);
-	D_FREE(ctx);
+	CTX_DECREF(ctx);
```

Comment:
I don't have a good idea, only an average one: Instead of calling `CTX_DEFREF(ctx)`, we wait for the reference count of `ctx` to drop to 1, then call `crt_context_free`.